### PR TITLE
Fix link issues in PDF build 4.2

### PR DESCRIPTION
**Pull request by rejahrehim**
- [x] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

**What did this PR accomplish?**

- Fix linking issue  in PDF while including links to the Appendix 
- Fix linking issue  in PDF while including external markdown URLs with fragment

### CODE REVIEW

**Review by kingthorin — APPROVED** [review]
Thanks

**Comment by ThunderSon** [comment]
Still reviewing the PDF.
For now:
1. The links in the sidebar are broken (if they are related to this script).
2. In the GraphQL **References** section, Client-side Testing has a reference to localhost (maybe it happened cause of the - in client-side. Page 435 of the PDF, 432 in the book numbering

**Comment by rejahrehim** [comment]
> Still reviewing the PDF.
> For now:
> 
>     1. The links in the sidebar are broken (if they are related to this script).
> 
>     2. In the GraphQL **References** section, Client-side Testing has a reference to localhost (maybe it happened cause of the - in client-side. Page 435 of the PDF, 432 in the book numbering

 I will update this PR with the fixes

**Comment by rejahrehim** [comment]
> > Still reviewing the PDF.
> > For now:
> > ```
> > 1. The links in the sidebar are broken (if they are related to this script).
> > 
> > 2. In the GraphQL **References** section, Client-side Testing has a reference to localhost (maybe it happened cause of the - in client-side. Page 435 of the PDF, 432 in the book numbering
> > ```
> 
> I will update this PR with the fixes

These two issues fixed. @ThunderSon @kingthorin

**Review by kingthorin — APPROVED** [review]
Thanks

**Comment by ThunderSon** [comment]
In the latest build, WSTG-CRYP-03 in the summary, the ASVS link isn't working properly. I am guessing other new github links don't work. I'll try and find more examples.

**Comment by rejahrehim** [comment]
> In the latest build, WSTG-CRYP-03 in the summary, the ASVS link isn't working properly. I am guessing other new github links don't work. I'll try and find more examples.

Fixed

**Review by ThunderSon — CHANGES_REQUESTED** [review]
On line 246, it should `ABCDEF` now with the new appendix :)
Fantastic work mate.

File: .github/pdf/scripts/make-pdf.sh [hunk 1]
```diff
@@ -1,185 +1,243 @@
...
+# Convert all chars inside id to lower case
+convert_id_to_lower_case () {
+    python -c "import re; import sys; print(re.sub(r'id=\"([^\n]+)\"', lambda m: m.group().lower(), sys.stdin.read()))"
+}
+
+# Remove `:`, `,`, `.` inside id values
+remove_special_chars_from_id () {
+    python -c "import re; import sys; print(re.sub(r'id=\"([^\n]+)\"', lambda m: m.group().replace(':', ''), sys.stdin.read()))"  | \
+    python -c "import re; import sys; print(re.sub(r'id=\"([^\n]+)\"', lambda m: m.group().replace('.', ''), sys.stdin.read()))"  | \
+    python -c "import re; import sys; print(re.sub(r'id=\"([^\n]+)\"', lambda m: m.group().replace(',', ''), sys.stdin.read()))"
```
[the remark this answers is not in the document]
Comment (rejahrehim) [reply .github/pdf/scripts/make-pdf.sh]: Checked. It works !

File: .github/pdf/scripts/make-pdf.sh [hunk 2]
```diff
@@ -1,185 +1,243 @@
...
+}
+
+# Replace the spaces and _ inside `href` values with hyphen
+# Handles the href with .md extention ( Keeps this .md inside the href to identify internal markdown links)
+replace_space_and_underscore_with_hyphen_in_internal_markdown_href () {
+
+    # Replace space
+    python -c "import re; import sys; print(re.sub(r'href=\"(#[^\"]*\.md)\"', lambda m: m.group().replace(' ', '-'), sys.stdin.read()))" | \
+    # Replace the `_`
+    python -c "import re; import sys; print(re.sub(r'href=\"(#[^\"]*\.md)\"', lambda m: m.group().replace('_', '-'), sys.stdin.read()))"
```
[the remark this answers is not in the document]
Comment (rejahrehim) [reply .github/pdf/scripts/make-pdf.sh]: Checked. It works !

**Comment by rejahrehim** [comment]
> On line 246, it should `ABCDEF` now with the new appendix :)
Fixed 👍

**Review by ThunderSon — APPROVED** [review]
Thanks!

**Pull request merged**